compiler/ballerina-compiler-api/src/main/java/org/ballerina/compiler/impl/BallerinaSemanticModel.java

```diff
@@ -102,15 +105,16 @@ public BallerinaSemanticModel(BLangPackage bLangPackage, CompilerContext context
      */
     @Override
     public Optional<Symbol> symbol(String srcFile, LinePosition position) {
-        SymbolEnv enclosingEnv = this.envResolver.lookUp(getCompilationUnit(srcFile), position);
-
-        for (Map.Entry<Name, Scope.ScopeEntry> entry : enclosingEnv.scope.entries.entrySet()) {
-            Name name = entry.getKey();
-            Scope.ScopeEntry value = entry.getValue();
-
-            if (isWithinSymbol(position, value.symbol.pos)) {
-                return Optional.of(SymbolFactory.getBCompiledSymbol(value.symbol, name.value));
-            }
+        BLangCompilationUnit compilationUnit = getCompilationUnit(srcFile);
+        NodeResolver nodeResolver = new NodeResolver();
+        BLangNode node = nodeResolver.lookup(compilationUnit, position);
+
+        if (node instanceof IdentifiableNode) {
```

Comment:
Can we check the node-kind and get the symbol?
Checking `instanceof` can be expensive, and we won't have to introduce another interface also.

Reply:
If checking the node kind, we'll have to handle each node with a symbol separately. What I have done here is, introduce an API for getting/setting a symbol to a node (currently we don't have a unified API for accessing a symbol from a node). So any node which has a symbol associated to it implements this interface and at this point, once I find the node enclosing the specified position, I just simply call `getSymbol()` to get the symbol. No need to consider the nodes individually.

Reply:
+1 go with this for the initial version. But IMO we should remove `instanceof` checks in all places.

For e.g.: since the NodeResolver is already a visitor, if we can return the symbol directly rather than doing it in two-steps, then we can avoid these checks.

Reply:
The performance of instanceof check isn't an issue in Java 8 right? 

Made it a generic node resolving visitor in case we need similar behaviours for other APIs. +1 for making it return the symbol itself according how this progresses